### DISCUSSION TITLE
Question whether to free message in process_message on error

### DISCUSSION THREAD

**Discussion by quinnj** [awslabs/aws-c-io#632] [Q&A] [answered]
I couldn't find anything [explicitly documented](https://github.com/awslabs/aws-c-io/blob/main/include/aws/io/channel.h#L55) as such, but I'm trying to confirm a pattern I found in code. For example, [here](https://github.com/awslabs/aws-c-io/blob/main/source/darwin/secure_transport_tls_channel_handler.c#L661), if the message sending fails for some reason, the handler itself will free the message. So in my channel handler vtable `process_read_message` implementation, I should only free the message once I've _successfully_ handle it, right? i.e. if I have an error condition for whatever reason, then I _shouldn't_ free the message.

I ask because I had originally done the equivalent of:
```
try
    internal_process_read_message(message)
finally
    aws_mem_release(alloc, message)
end
```
to ensure I always handled the message freeing, but then ran into rare edge cases where my `internal_process_read_message` implementation would throw (and thus return an error code from my vtable `process_read_message`) and then, I guess due to a race condition, _MY_ `aws_mem_release` on the message would fail an assertion because the sending channel handler would have already released it.

Am I understanding this all correctly? I could do a quick PR to the channel handler vtable docstring to clarify.

**Comment by waahm7** [comment, marked as answer]
@quinnj Thanks for creating the issue. Yes, you are right. You must only free the message if you have successfully handled it. Sorry for the confusing documentation. I have created a PR [https://github.com/awslabs/aws-c-io/pull/633](https://github.com/awslabs/aws-c-io/pull/633) to update it.
